File: WebSite/models/Business.ts
import mongoose from "mongoose";
import { Schema, models, model } from "mongoose";

export interface BusinessDocument extends mongoose.Document {
  userId: mongoose.Types.ObjectId;
  name: string;
  slug: string;
  primaryUrl: string;
  subUrls: string[];
  competitorUrls: string[];
  category: string;
  description?: string;
  // Extended business context (used by the Python audit service)
  targetKeywords: string[];
  servicesOrProducts: string[];
  uniqueSellingPoints: string[];
  targetAudience?: string;
  priceRange?: "budget" | "mid" | "premium";
  yearFounded?: number;
  certifications: string[];
  socialMediaUrls: string[];
  // Locality (persisted so re-audits keep the same classification)
  localityTier?: "global" | "national" | "hyper_local";
  city?: string;
  country?: string;
  neighborhood?: string;
  street?: string;
  region?: string;
  deletedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const BusinessSchema = new Schema<BusinessDocument>(
  {
    userId: { type: Schema.Types.ObjectId, ref: "User", required: true },
    name: { type: String, required: true, trim: true, maxlength: 100 },
    slug: { type: String, required: true, lowercase: true, trim: true },
    primaryUrl: { type: String, required: true, trim: true },
    subUrls: { type: [String], default: [] },
    competitorUrls: { type: [String], default: [] },
    category: { type: String, required: true, trim: true, maxlength: 100 },
    description: { type: String, maxlength: 500 },
    // Extended business context
    targetKeywords: { type: [String], default: [] },
    servicesOrProducts: { type: [String], default: [] },
    uniqueSellingPoints: { type: [String], default: [] },
    targetAudience: { type: String, trim: true, maxlength: 200 },
    priceRange: { type: String, enum: ["budget", "mid", "premium"] },
    yearFounded: { type: Number, min: 1800, max: 9999 },
    certifications: { type: [String], default: [] },
    socialMediaUrls: { type: [String], default: [] },
    // Locality
    localityTier: { type: String, enum: ["global", "national", "hyper_local"] },
    city: { type: String, trim: true, maxlength: 100 },
    country: { type: String, trim: true, maxlength: 100 },
    neighborhood: { type: String, trim: true, maxlength: 100 },
    street: { type: String, trim: true, maxlength: 200 },
    region: { type: String, trim: true, maxlength: 100 },
    deletedAt: { type: Date, default: null },
  },
  { timestamps: true }
);

// Compound index: unique slug per user
BusinessSchema.index({ userId: 1, slug: 1 }, { unique: true });
BusinessSchema.index({ userId: 1, deletedAt: 1 });

// Auto-generate slug from name before validation
BusinessSchema.pre("validate", function (next) {
  if (this.isModified("name") && !this.isModified("slug")) {
    this.slug = this.name
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-|-$/g, "")
      .substring(0, 60);
  }
  next();
});

const Business = (models?.Business || model<BusinessDocument>("Business", BusinessSchema)) as mongoose.Model<BusinessDocument>;
export default Business;

File: server/src/models/business.py
"""
Business metadata Pydantic models — flexible, type-agnostic.

AuditRequest accepts any business type. The prompt generator LLM receives
ALL fields and adapts prompts accordingly.
"""

from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, HttpUrl

from models.audit import PyObjectId


class LocalityTier(str, Enum):
    """How geographically anchored the business is."""

    HYPER_LOCAL = "hyper_local"   # coffee shop, restaurant, barber, gym
    NATIONAL = "national"         # German shoe brand, national delivery
    GLOBAL = "global"             # SaaS, dropshipping, Amazon-style


class AuditRequest(BaseModel):
    """Request model for the /audit endpoint."""

    model_config = ConfigDict(arbitrary_types_allowed=True)

    # Required fields — every audit must have these
    auditId: str  # MongoDB ObjectId string, created by Next.js
    businessId: PyObjectId
    userId: PyObjectId

    # Core business identity
    businessName: str  # Brand/business name (e.g., "MaisonCuir")
    businessUrl: str  # Primary website URL
    businessType: str  # Free-text type (e.g., "coffee-shop", "e-commerce", "saas")
    category: str  # Business category (e.g., "artisanal leather shoes")
    description: str  # Business description (2-5 sentences)
    language: str = "fr"  # Language for prompts ("fr" or "en")

    # Locality tier (auto-classified if not provided)
    localityTier: LocalityTier | None = None

    # Location (optional — for local businesses)
    city: str | None = None
    neighborhood: str | None = None
    street: str | None = None
    country: str | None = None
    region: str | None = None

    # Business details (optional — enriches prompt quality)
    subUrls: list[str] = []  # Sub-pages to analyze (e.g., /blog, /shop)
    competitorUrls: list[str] = []  # Competitor URLs (max 5)
    competitorNames: list[str] = []  # Competitor brand names
    targetKeywords: list[str] = []  # SEO target keywords
    uniqueSellingPoints: list[str] = []  # What makes the business special
    targetAudience: str | None = None  # Who the business serves
    priceRange: str | None = None  # Budget/mid/premium
    yearFounded: int | None = None  # When business was established
    servicesOrProducts: list[str] = []  # Main offerings
    certifications: list[str] = []  # Organic, fair-trade, ISO, etc.
    socialMediaUrls: list[str] = []  # Social presence

    # Catch-all for business-specific data
    customFields: dict = {}

    # Google Search Console (optional — frontend provides OAuth2 token)
    gscAccessToken: str | None = None   # GSC OAuth2 access token
    gscSiteUrl: str | None = None       # GSC property URL (may differ from businessUrl)

    googlePlaceId: str | None = None  # Auto-detected if not provided


class GscAnalysisRequest(BaseModel):
    """Request model for the standalone /gsc-analysis endpoint."""

    accessToken: str
    siteUrl: str       # GSC property (e.g. "https://example.com/" or "sc-domain:example.com")
    primaryUrl: str    # URL to inspect via the URL Inspection API
    localityTier: str | None = None   # For geo relevance scoring
    country: str | None = None        # Business's expected country (ISO alpha-2)
    daysBack: int = 28


class HtmlScanRequest(BaseModel):
    """Request model for the /html-scan endpoint."""

    url: HttpUrl = Field(..., max_length=2048)
    subUrls: list[str] = Field(default=[], max_length=10)
    language: str = "fr"
    businessType: str | None = None  # e.g. "e-commerce", "restaurant", "news", "saas"
